x/protorev/keeper/posthandler.go

```diff
@@ -0,0 +1,123 @@
+package keeper
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	gammtypes "github.com/osmosis-labs/osmosis/v13/x/gamm/types"
+)
+
+type SwapToBackrun struct {
+	PoolId        uint64
+	TokenOutDenom string
+	TokenInDenom  string
+}
+
+type ProtoRevDecorator struct {
+	ProtoRevKeeper Keeper
+}
+
+func NewProtoRevDecorator(protoRevDecorator Keeper) ProtoRevDecorator {
+	return ProtoRevDecorator{
+		ProtoRevKeeper: protoRevDecorator,
+	}
+}
+
+// This posthandler will first check if there were any swaps in the tx. If so, collect all of the pools, build three
+// pool routes for cyclic arbitrage, and then execute the optimal route if it exists.
+func (protoRevDec ProtoRevDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (sdk.Context, error) {
+	// Create a cache context to execute the posthandler such that
+	// 1. If there is an error, then the cache context is discarded
+	// 2. If there is no error, then the cache context is written to the main context
+	cacheCtx, write := ctx.CacheContext()
+
+	txGasWanted := cacheCtx.GasMeter().Limit()
+	// Ignore cases where limit is 0 (edge case for genUtil init)
+	if txGasWanted == 0 {
+		return next(ctx, tx, simulate)
+	}
+
+	// Change the gas meter to be an infinite gas meter which allows the entire posthandler to work without consuming any gas
+	cacheCtx = cacheCtx.WithGasMeter(sdk.NewInfiniteGasMeter())
+
+	// Only execute the protoRev module if it is enabled
+	if enabled, err := protoRevDec.ProtoRevKeeper.GetProtoRevEnabled(cacheCtx); err != nil || !enabled {
+		return next(ctx, tx, simulate)
+	}
+
+	// Get the max number of pools to iterate through
+	maxPoolsToIterate, err := protoRevDec.ProtoRevKeeper.GetMaxPools(cacheCtx)
+	if err != nil {
+		return next(ctx, tx, simulate)
+	}
+
+	// Find routes for every single pool that was swapped on (up to maxPoolsToIterate pools per tx)
+	swappedPools := ExtractSwappedPools(tx)
+	tradeErr := error(nil)
+	for index, swap := range swappedPools {
```

Comment:
nit: I think it's cleaner to take the min of len(swappedPools) and maxPoolsToIterate and just iterate over that, so you can get rid of a break statement